source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -492,16 +493,18 @@ core::smart_refctd_ptr<IGPUSpecializedShader> COpenGLDriver::createGPUSpecialize
         if (!spvShader)
             return nullptr;
 
-        spv = spvShader->getSPVorGLSL();
-        spv->grab();
+        spvCPUShader = spvShader.get();
+        spvCPUShader->grab();
     }
     else {
-        spv = cpuUnspec->getSPVorGLSL();
-        spv->grab();
+        spvCPUShader = cpuUnspec;
+        spvCPUShader->grab();
     }
 
-    COpenGLSpecializedShader* gpuSpecialized = new COpenGLSpecializedShader(this, spv, _specInfo);
-    spv->drop();
+    asset::CShaderIntrospector introspector(GLSLCompiler.get(), {_specInfo->entryPoint, _specInfo->shaderStage});
+    const asset::CIntrospectionData* introspection = introspector.introspect(spvCPUShader);
+    COpenGLSpecializedShader* gpuSpecialized = new COpenGLSpecializedShader(this, spvCPUShader->getSPVorGLSL(), _specInfo, introspection);
+    spvCPUShader->drop();
```

Comment:
I don't want to see grab and drops in the engine anymore

Reply:
@Crisspl ?

Reply:
this drop is needed here

Reply:
why arent you dealing with smart pointers in the first place?

Reply:
read the code and u'll know why im doing this

Reply:
gonna mark this resolved, because its  duplicate of the https://github.com/buildaworldnet/IrrlichtBAW/pull/358#discussion_r324890213 
thread.

Reply:
resolved